config/initializers/heinous_inline_partial.rb

```diff
@@ -0,0 +1,61 @@
+# Inlines partials into parent templates for performance.
+# In production & test: once at startup; development: every request
+#
+# Inlining the comments/_comment partial into stories/show (bulk of our traffic!)
+# gives a 25-50% speedup over 'render collection: @comment', and 80-90% speedup over looping to
+# 'render partial: comment'.
+
+# rubocop:disable Style/MutableConstant
+HEINOUS_INLINE_PARTIALS = {
```

Comment:
I'm wondering if we could go for a `render collection` with some rails cache tweaks to get a "good enough" speedup to skip this. While the performance is great, I'm concerned as to how well its going to upgrade.

Reply:
I share your preference for caching, and the logged-out users get a cache from `StoriesController#get_arranged_comments_from_cache`. But comments include per-user vote arrows, so we'd need to cache per-user, and I'd rather logged-in-users see discussions live. The recursive CTE takes fetching + arranging comments from ~30ms to ~5ms, so there's not a lot of win available on that side - the expensive part is the rendering.

Inlining the partial is roughly a 2x speedup over `render @collection` ([https://medium.com/@coorasse/partial-rendering-performance-in-rails-101fdfb6ffb9](seems a common result)). So, with full awareness that I am doing something kind of awful, I tried to reduce the likelihood of footguns - prevent recursion or chaining, require conspicuous templating literals, emit conspicuous warnings at startup.

--

As long as I am braindumping: the part of the code I am least satisfied with is in `app/views/comment/_thread.html.erb`, which has a lot of clever and unstructured string munging to nest comments properly in a single pass. A comment has to be rendered without knowing if it has children so 1. the `.comment_parent_tree_line` becomes brittle css, and 2. the `<ol>` has to be inferred when rendering the child and `</ol>` when rendering the next comment at equal or lesser depth, or end of the comment tree. And because a comment has to be rendered without knowing its siblings, the `.comment_sibling_tree_line` is also a clever bit of css, though at least it's not brittle. (The brittle parent line could be a short + reliable use of `:has`, which currently is almost universally supported - it's [in testing](https://caniuse.com/css-has) in Firefox.)

All this ugliness could be fixed by adding a `reply_count` [counter cache](https://guides.rubyonrails.org/association_basics.html#options-for-belongs-to-counter-cache) to `comments` to track the number of direct replies to each comment. That solves the first two problems of knowing if a comment has children when rendering it are solved, and adding to a hash like `reply_counts[comment.id] = comment.reply_count` as comments are rendered would let children look up `reply_counts[comment.parent_comment_id]` to know if they have siblings.

I think the current solution is acceptable as long as it's not breaking, and I think we'll touch it infrequently enough that this is acceptable. I am significantly dissatisfied that I couldn't find a working HTML validator gem to make sure we are outputing properly nested and closed `<ol>` and `<li>`s; I tested by hand-pasting into the [w3c validator](https://validator.w3.org/#validate_by_input). If this becomes a problem, I'd start to think it was worth the cost of widening `comments` even more.